utilities/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+    let path_dat = paths.path_dat.as_path();
+    let modified_out = match path_dat.metadata() {
+        Ok(meta) => meta.modified().expect("Failed to read dat file metadata"),
+        Err(_error) => {
+            fs::create_dir_all(path_dat.parent().expect("Failed to create dat folder"))?;
+            time::SystemTime::UNIX_EPOCH
+        }
+    };
+    if modified_out < modified_in {
+        // The unzipped version either doesn't exist, or is older than the zipped version.
+        let file_in = fs::File::open(path_zip).expect("Failed to open zip file");
+        let mut zip_archive = zip::ZipArchive::new(file_in).expect("Failed to read zip archive");
+        if zip_archive.len() != 1 {
+            panic!("Expected zip archive to contain exactly one file");
+        }
+        let mut zip_first = zip_archive.by_index(0).expect("Failed to open first file in zip archive");
+        let mut zip_first_contents = String::new();
+        zip_first.read_to_string(&mut zip_first_contents).expect("Failed to read zip file content");
+        fs::write(path_dat, zip_first_contents).expect("Failed to write zip content to dat file");
+    }
+    Ok(())
+}
+
+pub fn as_f64(s: &str) -> Result<f64, Box<dyn Error + Sync + Send>> {
+    match s {
+        "nan" => Ok(f64::NAN),
+        "-nan" => Ok(-f64::NAN),
+        "inf" => Ok(f64::INFINITY),
+        "-inf" => Ok(f64::NEG_INFINITY),
+        _ => {
+            if s.starts_with("0x") {
+                // Bit fields are recorded using hex notation, to avoid endian difference complications.
+                let without_prefix = s.trim_start_matches("0x");
+                match u64::from_str_radix(without_prefix, 16) {
+                    Ok(num) => Ok(num as f64),
+                    Err(err) => Err(Box::<dyn Error + Sync + Send>::from(err.to_string())),
+                }
+            } else {
+                // All other value types are recorded using values that can be parsed as f64.
+                // In some cases, they'll need to be coerced into a different type later.
+                match s.parse::<f64>() {
+                    Ok(num) => Ok(num),
+                    Err(err) => Err(Box::<dyn Error + Sync + Send>::from(err.to_string())),
+                }
+            }
+        },
+    }
+}
+
+// Return the absolute and relative difference between two values, in that order.
+// If both values are nan, consider the difference to be 0.
+pub fn calc_delta(x: f64, y: f64) -> (f64, f64) {
+    if x.is_nan() && y.is_nan() {
+        return (0f64, 0f64);
+    } else if x.is_infinite() && y.is_infinite() {
+        return if x.is_sign_negative() == y.is_sign_negative() { (0f64, 0f64) } else { (f64::INFINITY, f64::INFINITY) };
+    } else {
+        let diff_abs = (x - y).abs();
+        let diff_rel = 2.0 * diff_abs / (x.abs() + y.abs());
+        return (diff_abs, diff_rel);
+    }
+}
+
+// Extract a the first value from each line, and return as a vector.
```

Comment:
By "extract the first value" you mean get rid of it right?

Reply:
No, it's referring to fetching the first value of each line, and ignoring the rest. Data lines are structured to list inputs first, and then outputs, even if the function uses a more mixed order. For a benchmark, we don't even care what the outputs are, so I use either as_vecs_basic or as_vec_basic to extract just the inputs. I decided there were enough operations that use a single input that it was worth having as_vec_basic for that special case, where it returns a vector of just the first value on each line, rather than a vector of vectors.